vulcanai/models/basenetwork.py

```diff
@@ -550,6 +651,23 @@ def fit(self, train_loader, val_loader, epochs,
                 if self.lr_scheduler:
                     self.lr_scheduler.step(epoch=epoch)
 
+                if self.early_stopping:
+                    if self.early_stopping_metric == "loss":
+                        early_stopping(-1* valid_loss, self)
+                    elif self.early_stopping_metric == "accuracy":
+                        early_stopping(valid_acc, self)
+                    else:
+                        raise ValueError("Not a valid stopping metric")
+
+                    if early_stopping.early_stop:
+                        logger.info("Early stopping")
+                        # restores everything to the earlier saved version
+                        self.__dict__.update(self.load_model(
```

Comment:
is there some test we can write that would capture this change works as intended? perhaps a very small toy model that deterministically learns the same way every time and we can trigger such event @nirtiac ?

Reply:
yeah but that doesn't really test if we've kept the correct one. but. sure.

Reply:
True - In that case how about we keep track of the best model outside the training process and then take a hash of the model. Then once the training is complete you can do a simple checksum of the model that was self selected and see if it matches exactly the one that the external method found as well? Shouldn’t be too demanding if it’s a small toy model

Reply:
had thought of that, but https://discuss.pytorch.org/t/why-does-the-hash-value-of-models-change-everytime-i-save-it/24249

Reply:
so would only work if it's truly the same object. I tested it by comparing all parameters (manually), could do something similar.

Reply:
Ah yes - very good point. It makes sense to reserve the checks for just the parameters

Reply:
actually, no can do, as we don't have control or access to the model at the time when the best model is found.

Reply:
perhaps if we set up a hook and trigger a pull under some condition we can access it? similar to how the GuidedBackprop works?

Reply:
I think that's too much to set up just for a test imo

Reply:
Fair enough. Is there some other alternative that you think we can test with this? What if we train two separate but identically initialized models and see if the final versions are the same?